model/src/main/java/com/example/opencommerce/domain/pricing/ItemPrice.java

```diff
@@ -0,0 +1,185 @@
+        return executePriceChange(newPrice, executionDate);
+    }
+
+    public OperationResult<DiscountAppliedEvent> addNewDiscount(Value discount, Instant executionDate) {
+        return executeDiscount(discount, executionDate);
+    }
+
+    public OperationResult<DiscountRemovedEvent> removeCurrentDiscount(Instant executionDate) {
+        return executeDiscountRemoval(executionDate);
+    }
+
+    private Value lowestPriceWithinPeriod(Period timePeriod) {
+        Instant startDate = Instant.now()
+                .minus(timePeriod);
+        return this.history.stream()
+                .filter(price -> price.getValidTo()
+                        .isAfter(startDate))
+                .min(Comparator.comparing(p -> p.getPriceNett()
+                        .asDecimal()))
+                .map(HistoricalPrice::getPriceNett)
+                .orElse(this.basePriceNett);
+    }
+
+    public PriceSnapshot getSnapshot() {
+        if (discounted()) {
+            return PriceSnapshot.discountedPrice(this.id, this.basePriceNett, this.discount, lowestPriceWithinPeriod(Period.ofDays(30)), this.vat);
+        } else {
+            return PriceSnapshot.regularPrice(this.id, this.basePriceNett, this.vat);
+        }
+    }
+
+    private void apply(PriceEvent e) {
+
+        OperationResult<?> result;
+        if (e instanceof NewPriceInitiatedEvent event) {
+            result = executePriceInitiate(Value.of(event.getBasePrice()), Vat.of(event.getVat()));
+        } else if (e instanceof BasePriceChangedEvent event) {
+            result = executePriceChange(Value.of(event.getNewPrice()), event.getValidFrom());
+        } else if (e instanceof DiscountAppliedEvent event) {
+            result = executeDiscount(Value.of(event.getDiscount()), event.getValidFrom());
+        } else if (e instanceof DiscountRemovedEvent event) {
+            result = executeDiscountRemoval(event.getValidFrom());
+        } else {
+            throw new IllegalStateException("Unknown price event type: " + e.getClass()
+                    .getSimpleName());
+        }
+
+        if (!result.successful()) {
+            throw new IllegalStateException(
+                    format("Critical error during aggregate recovery! Problems: \n%s", String.join("\n", result.getErrors())));
+        }
+    }
+
+    private OperationResult<NewPriceInitiatedEvent> executePriceInitiate(Value priceNett, Vat vat) {
+        if (isInitiated()) {
+            return OperationResult.failed("Can't perform initiation on already initiated price record!");
+        }
+
+        this.basePriceNett = priceNett;
+        this.vat = vat;
+        this.initiated = true;
+
+        NewPriceInitiatedEvent event = new NewPriceInitiatedEvent(priceNett, vat);
+        return OperationResult.success(event);
+    }
+
+    private OperationResult<BasePriceChangedEvent> executePriceChange(Value newPrice, Instant executionTime) {
```

Comment:
The `execute price change` and `execute discount` common parts (validation) could be extracted to a common method, leaving the business logic methods cleaner, and simpler